### PR TITLE
[flutter_tool] Have long-running validators fail

### PR DESCRIPTION
**Pull request by christopherfujino**
Add debugging for https://github.com/flutter/flutter/issues/100731

### CODE REVIEW

File: packages/flutter_tools/lib/src/doctor.dart [hunk 1]
```diff
@@ -204,13 +207,29 @@ class Doctor {
...
+            final Completer<ValidationResult> timeoutCompleter = Completer<ValidationResult>();
+            final Timer timer = Timer(doctorDuration, () {
+              timeoutCompleter.completeError(
+                Exception('${validator.title} exceeded maximum allowed duration of $doctorDuration'),
+              );
+            });
+            final Future<ValidationResult> validatorFuture = validator.validate();
+            return Future.any<ValidationResult>(<Future<ValidationResult>>[
+              validatorFuture,
+              // This future can only complete with a ToolExit.
```
Comment (goderbauer) [inline packages/flutter_tools/lib/src/doctor.dart:220]: Looks like it just completes with a regular Exception rather then a ToolExit?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/doctor.dart (hunk 1 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/doctor.dart:220]: good catch

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/doctor.dart:220]: updated comment

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/doctor.dart]: (the exception type doesn't really matter, we catch them all on line 227, all that matters is the message)

**Comment by christopherfujino** [comment]
I'd like to see CI finish before merging this one

**Review by jmagman — APPROVED** [review]

**Comment by jmagman** [comment]
Failed test hit https://github.com/flutter/flutter/issues/100245

**Comment by christopherfujino** [comment]
merging on red tree to help fix the tree

**Pull request merged**